packages/core/src/dialects/mssql/query-generator-internal.ts

```diff
@@ -45,4 +51,67 @@ export class MsSqlQueryGeneratorInternal<Dialect extends MssqlDialect = MssqlDia
 
     return fragment;
   }
+
+  getReturnFields(options: GetReturnFieldsOptions, modelDefinition?: ModelDefinition | null): string[] {
+    const returnFields: string[] = [];
+
+    if (Array.isArray(options.returning)) {
+      returnFields.push(...options.returning.map(field => {
+        if (typeof field === 'string') {
+          return this.queryGenerator.quoteIdentifier(field);
+        } else if (field instanceof Literal) {
+          // Due to how the mssql query is built, using a literal would never result in a properly formed query.
+          // It's better to warn early.
+          throw new Error(`literal() cannot be used in the "returning" option array in ${this.dialect.name}. Use col(), or a string instead.`);
+        } else if (field instanceof BaseSqlExpression) {
+          return this.queryGenerator.formatSqlExpression(field, options);
+        }
+
+        throw new Error(`Unsupported value in "returning" option: ${inspect(field)}. This option only accepts true, false, an array of strings, or a sql expression.`);
+      }));
+    } else if (modelDefinition) {
+      for (const [name, attribute] of modelDefinition.physicalAttributes) {
+        returnFields.push(this.queryGenerator.quoteIdentifier(attribute.columnName ?? name));
+      }
+    }
+
+    if (returnFields.length === 0) {
+      returnFields.push('*');
+    }
+
+    return returnFields;
+  }
+
+  generateOutputTableFragment(returnFields: string[], modelDefinition?: ModelDefinition | null): string {
+    if (!(modelDefinition instanceof ModelDefinition)) {
+      throw new Error('Using returning with triggers requires specifying a model or model definition.');
```

Comment:
This check is specific to where this method was used. `modelDefinition` should be made mandatory here, and this check should be moved to where this method is called

Reply:
Updated now.